pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AbstractNode.java

```diff
@@ -229,6 +229,8 @@ public GenericToken jjtGetFirstToken() {
 
     public void jjtSetFirstToken(final GenericToken token) {
         this.firstToken = token;
+        this.beginLine = token.getBeginLine();
+        this.beginColumn = token.getBeginColumn();
```

Comment:
maybe just deprecate + remove the `begin*` and `end*` attributes and delegate always to the token?

Reply:
I think yes. I'm not sure this will be supported by all implementations though. I'll try it

Reply:
Wait so you mean to move those changes up into AbstractNode, right?

Reply:
Please in a separate PR...

Reply:
I'll add a TODO on https://github.com/pmd/pmd/wiki/PMD-7.0.0-Java